pcdsdevices/signal.py

```diff
@@ -266,6 +476,169 @@ def put(self, value, **kwargs):
+
+            self.calculate_on_put = utils.maybe_make_method(
+                calculate_on_put, self
+            )
+        elif type(self) is MultiDerivedSignal:
+            self._metadata["write_access"] = False
+            self.calculate_on_put = None
+
+        self.attrs = list(attrs)
+        if len(self.attrs) == 0:
+            raise ValueError(
+                "At least one attribute name must be supplied for a "
+                "MultiDerivedSignal.  Also consider using DerivedSignal "
+                "directly for the one-to-one case."
+            )
+
+        for attr_name in self.attrs:
+            self.add_signal_by_attr_name(attr_name)
+
+    @property
+    def signals(self) -> List[Signal]:
+        """The signals used to calculate this MultiDerivedSignal."""
+        return list(self._signals)
+
+    def _calc_readback(self) -> OphydDataType:
+        """Calculate the new readback value."""
+        cache = {sig: siginfo.value for sig, siginfo in self._signals.items()}
+        return self.calculate(cache)
+
+    def put(
+        self,
+        value: OphydDataType,
+        callback: Optional[OphydCallback] = None,
+        timeout: Union[Number, object] = DEFAULT_WRITE_TIMEOUT,
+        settle_time: Optional[float] = None,
+        **kwargs
+    ) -> ophyd.status.StatusBase:
+        """
+        Calculate new values for the given derived signals and write.
+
+        Keyword arguments outside of the listed parameters are ignored for
+        ophyd ``Signal.put()`` compatibility.
+
+        Parameters
+        ----------
+        value : OphydDataType
+            The value to set.  This is sent to the ``calculate_on_put``
+            function to determine which values to write to which signals.
+        callback : callable
+            Callback for when the put has completed.
+        timeout : float, optional
+            Timeout before assuming that put has failed.
+        settle_time : float, optional
+            Time after writing to wait before indicating the put has completely
+            finished.
+        """
+        if timeout is DEFAULT_WRITE_TIMEOUT:
+            timeout = self.timeout
+        if settle_time is DEFAULT_WRITE_TIMEOUT:
+            settle_time = self.settle_time
+        st = self.set(value, timeout=timeout, settle_time=settle_time)
+        if callback is not None:
+            st.add_callback(callback)
+        return st
+
+    def set(
+        self,
+        value: OphydDataType,
+        *,
+        timeout: Optional[float] = None,
+        settle_time: Optional[float] = None
+    ) -> ophyd.status.StatusBase:
+        if self.calculate_on_put is None:
+            raise ReadOnlyError(
+                f"{self.name} is read-only. "
+                f"`calculate_on_put` function not defined for {self.name} "
+                f"({type(self).__name__})."
+            )
+        to_write = self.calculate_on_put(value) or {}
+        if not isinstance(to_write, Mapping):
```

Comment:
I think technically the only acceptable result would be duck-typed bluesky `Movable` - https://github.com/bluesky/bluesky/blob/47e33af7f6d2c836ace4e54e2cf039a93c8b8beb/bluesky/protocols.py#L119 - but we can be a bit more restrictive here because we're working within the bounds of ophyd/pcdsdevices and not throwing other libraries into the mix.

Why not `isinstance(to_write, SignalToValue)`? Limitations of the type hinting system and being lazy about validation:
```python
In [1]: from pcdsdevices.type_hints import SignalToValue

In [2]: isinstance({}, SignalToValue)
...
TypeError: Subscripted generics cannot be used with class and instance checks
```